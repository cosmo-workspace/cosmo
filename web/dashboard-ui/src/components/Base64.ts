// https://github.com/google/webauthndemo/blob/main/src/public/scripts/base64url.ts

const chars =
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

// Use a lookup table to find the index.
const lookup = new Uint8Array(256);
for (let i = 0; i < chars.length; i++) {
  lookup[chars.charCodeAt(i)] = i;
}

const encode = function (arraybuffer: ArrayBuffer): string {
  const bytes = new Uint8Array(arraybuffer);
  const len = bytes.length;
  let base64 = "";

  for (let i = 0; i < len; i += 3) {
    base64 += chars[bytes[i] >> 2];
    base64 += chars[((bytes[i] & 3) << 4) | (bytes[i + 1] >> 4)];
    base64 += chars[((bytes[i + 1] & 15) << 2) | (bytes[i + 2] >> 6)];
    base64 += chars[bytes[i + 2] & 63];
  }

  if (len % 3 === 2) {
    base64 = base64.substring(0, base64.length - 1);
  } else if (len % 3 === 1) {
    base64 = base64.substring(0, base64.length - 2);
  }

  return base64;
};

const decode = function (base64: string): ArrayBuffer {
  const len = base64.length;
  const bufferLength = base64.length * 0.75;
  const arraybuffer = new ArrayBuffer(bufferLength);
  const bytes = new Uint8Array(arraybuffer);

  let p = 0;
  for (let i = 0; i < len; i += 4) {
    const encoded1 = lookup[base64.charCodeAt(i)];
    const encoded2 = lookup[base64.charCodeAt(i + 1)];
    const encoded3 = lookup[base64.charCodeAt(i + 2)];
    const encoded4 = lookup[base64.charCodeAt(i + 3)];

    bytes[p++] = (encoded1 << 2) | (encoded2 >> 4);
    bytes[p++] = ((encoded2 & 15) << 4) | (encoded3 >> 2);
    bytes[p++] = ((encoded3 & 3) << 6) | (encoded4 & 63);
  }

  return arraybuffer;
};

const base64url = { encode, decode };
export { base64url };
