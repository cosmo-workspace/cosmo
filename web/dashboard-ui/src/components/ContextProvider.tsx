import { createContext, ReactNode, useContext, useState } from "react";

/**
 * DialogContext
 */
type DialogState<T> = { open: boolean, dialogProps?: T }
type DialogProps<T> = T & { onClose: () => void }
type Dispatch<T> = (open: boolean, dialogProp?: T) => void;

export function DialogContext<T>(dialog: (props: DialogProps<T>) => any) {

  const Context = createContext<{ state: DialogState<T>, dispatch: Dispatch<T> }>(undefined as any);

  return {
    Provider: ({ children }: { children: ReactNode }) => {
      const [state, setState] = useState<DialogState<T>>({ open: false });

      const dispatch: Dispatch<T> = (open, dialogProps?) => {
        setState({ open, dialogProps });
      }
      const closeHandler = () => dispatch(false, state.dialogProps);
      return (
        <Context.Provider value={{ state, dispatch }}><>
          {children}
          {state.open && dialog({ ...state.dialogProps, onClose: () => closeHandler() } as DialogProps<T>)}
        </></Context.Provider>
      );
    },
    useDispatch: () => useContext(Context).dispatch,
  }
}

/**
 * ModuleContext
 */
export function ModuleContext<T extends () => any>(useModule: T) {
  const Context = createContext<ReturnType<T>>(undefined as any);
  return {
    Provider: ({ children }: { children: ReactNode }) => {
      const module = useModule();
      return (<Context.Provider value={{ ...module }}>{children}</Context.Provider>);
    },
    useContext: () => useContext(Context),
  }
}
