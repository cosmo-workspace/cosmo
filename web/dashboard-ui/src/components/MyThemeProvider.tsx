import { colors, useMediaQuery } from "@mui/material";
import { createTheme, ThemeProvider } from "@mui/material/styles";
import React, { useMemo } from "react";

const MyTheme = () => {
  console.log('MyTheme');
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)', { noSsr: true });

  return useMemo(() => {

    return createTheme({
      components: {
        MuiOutlinedInput: {
          defaultProps: {
            notched: true,
          },
        },
        MuiInputLabel: {
          defaultProps: {
            shrink: true,
          },
        },
      },
      palette: {
        mode: prefersDarkMode ? 'dark' : undefined,
        primary: colors.deepPurple,
        secondary: colors.pink,
      },
    });
  }, [prefersDarkMode]);
}

export const MyThemeProvider: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => {

  const myTheme = MyTheme();

  return (
    <ThemeProvider theme={myTheme}>
      {children}
    </ThemeProvider>
  );
};
