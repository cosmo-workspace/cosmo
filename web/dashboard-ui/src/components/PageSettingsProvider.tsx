import React, { createContext, useContext, useEffect, useReducer } from 'react';

interface PageSettings {
  isOpen: boolean;
};

interface PageSettingsContext {
  pageSettings: PageSettings;
  setPageSettings: (setting: PageSettings) => void;
  updatePageSettings: (setting: PageSettings) => void;
};

const Context = createContext<PageSettingsContext>({
  pageSettings: {
    isOpen: true,
  },
  setPageSettings: (_setting) => { },
  updatePageSettings: (_setting) => { },
});

interface Action {
  type: 'SET' | 'UPDATE';
  pageSettings: PageSettings;
}
function reducer(state: PageSettings, action: Action) {
  switch (action.type) {
    case 'SET':
      return action.pageSettings;
    case 'UPDATE':
      return { ...state, ...action.pageSettings }
    default:
      throw new Error()
  }
}

export const PageSettingsProvider: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => {
  const persistKey = 'csm_PageSettings';
  const persistPageSettings = JSON.parse(localStorage.getItem('csm_PageSettings') || "{}");
  const [pageSettings, dispatch] = useReducer(reducer, persistPageSettings || {})

  useEffect(() => {
    try {
      localStorage.setItem(persistKey, JSON.stringify(pageSettings))
    } catch (error) {
      console.warn(error)
    }
  }, [pageSettings])

  return (
    <Context.Provider value={{
      pageSettings,
      setPageSettings: (settings) => { dispatch({ type: 'SET', pageSettings: settings }) },
      updatePageSettings: (settings) => { dispatch({ type: 'UPDATE', pageSettings: settings }) }
    }}>
      {children}
    </Context.Provider>
  )
}

export function usePageSettings() {
  return useContext(Context);
}
