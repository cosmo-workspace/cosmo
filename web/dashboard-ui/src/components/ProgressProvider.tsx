import { Backdrop, CircularProgress } from '@mui/material';
import React, { createContext, useContext, useState } from 'react';

const DispatchContext = createContext<React.Dispatch<React.SetStateAction<number>>>(undefined as any);

/**
 * provider
 */
export const ProgressProvider: React.FC = ({ children }) => {
  const [count, setCount] = useState(0);
  return (<>
    <DispatchContext.Provider value={setCount}>
      {children}
    </DispatchContext.Provider>
    <Backdrop sx={{ zIndex: (theme) => theme.zIndex.drawer + 1000 }} open={count > 0}>
      <CircularProgress />
    </Backdrop>
  </>);
}

export function useProgress() {
  const setCount = useContext(DispatchContext);
  return {
    setMask: () => setCount(count => { console.log('setMask', count + 1); return count + 1 }),
    releaseMask: () => setCount(count => { console.log('releaseMask', count); return count > 0 ? count - 1 : 0 }),
  }
}