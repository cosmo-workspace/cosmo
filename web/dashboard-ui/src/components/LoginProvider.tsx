import { Box, CircularProgress, CssBaseline, Stack, Typography } from '@mui/material';
import { useSnackbar } from 'notistack';
import React, { createContext, useContext, useEffect, useState } from 'react';
import { User } from '../proto/gen/dashboard/v1alpha1/user_pb';
import { useAuthService, useUserService } from '../services/DashboardServices';
import { useProgress } from './ProgressProvider';


/**
 * context
 * ref: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24509#issuecomment-774430643
 */
const Context = createContext<ReturnType<typeof useLoginModule>>(undefined as any);

/**
 * hooks
 */
export function useLogin() {
  return useContext(Context);
}

const useLoginModule = () => {
  console.log('useLoginModule');
  const [loginUser, setLoginUser] = useState<User>();
  const { enqueueSnackbar } = useSnackbar();
  const { setMask, releaseMask } = useProgress();
  const userService = useUserService();
  const authService = useAuthService();


  /**
   * verifyLogin
   */
  const verifyLogin = async () => {
    console.log('verify start');
    try {
      const resp = await authService.verify({});
      if (resp.userName) {
        await getMyUserInfo(resp.userName);
      }
    }
    catch (error) {
      setLoginUser(undefined);
      //handleError(error);
    }
    finally {
      console.log('verify end');
    }
  }


  /**
   * login: SignIn 
   */
  const login = async (userName: string, password: string) => {
    console.log('login');
    try {
      const res = await authService.login({ userName: userName, password: password });
      await getMyUserInfo(userName);
      console.log('login end');
      return res;
    }
    catch (error) {
      setLoginUser(undefined);
      console.log('login error');
      handleError(error);
      throw error;
    }
  }


  /**
   * login: MyUserInfo
   */
  const getMyUserInfo = async (userName: string) => {
    console.log('getMyUserInfo', userName);
    // if (loginUser || !username) {
    //   console.log('getMyUserInfo cancel', loginUser, username);
    //   return;
    // }
    try {
      const responseUser = await userService.getUser({ userName: userName });
      setLoginUser(prev => { console.log('setLoginUser', prev, responseUser.user); return responseUser.user! });
    }
    catch (error) {
      console.log('getMyUserInfo error');
      handleError(error);
      throw error;
    }
  }

  /**
   * login: MyUserInfo
   */
  const refreshUserInfo = async () => {
    console.log('refreshUserInfo');
    if (loginUser) {
      getMyUserInfo(loginUser.name);
    }
  }

  /**
   * logout: Pagetemplate 
   */
  const logout = async () => {
    console.log('logout');
    try {
      await authService.logout({});
    }
    catch (error) {
      console.log('logout error');
      handleError(error);
      throw error;
    }
    finally {
      console.log('logout end');
      setLoginUser(prev => { console.log('setLoginUser', `${prev} -> undefined`); return undefined });
    }
  }

  /**
   * updataPassword
   */
  const updataPassword = async (currentPassword: string, newPassword: string) => {
    console.log('updataPassword', loginUser?.name);
    setMask();
    try {
      try {
        return await userService.updateUserPassword({ userName: loginUser!.name, currentPassword, newPassword });
      }
      catch (error) {
        handleError(error);
        throw error;
      }
    }
    finally {
      console.log('updataPassword end');
      releaseMask();
    }
  }

  /**
   * clearLoginUser 
   */
  const clearLoginUser = async () => {
    setLoginUser(undefined);
  }

  /**
   * error handler
   */
  const handleError = (error: any) => {
    console.log(error);
    const msg = error?.message;
    msg && enqueueSnackbar(msg, { variant: 'error' });
  }

  return {
    loginUser,
    verifyLogin,
    login,
    logout,
    refreshUserInfo,
    updataPassword,
    clearLoginUser,
  };
}

/**
 * Provider
 */
export const LoginProvider: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => {
  console.log('LoginProvider');
  const loginModule = useLoginModule();
  const [isVerified, setIsVerified] = useState(false);

  useEffect(() => {
    loginModule.verifyLogin()
      .then(() => setIsVerified(true));
  }, []); // eslint-disable-line

  return (
    <Context.Provider value={loginModule}>
      {isVerified ? children :
        <div>
          <CssBaseline />
          <Stack sx={{ m: 10 }} alignItems='center' spacing={2}>
            <Box sx={{ display: 'flex' }}>
              <CircularProgress />
            </Box>
            <Typography >verifying session...</Typography>
          </Stack>
        </div>}
    </Context.Provider >
  )
}

/**
 * export private member. (for test)
 */
// export const __local__ = {
//   useLoginModule,
// };
