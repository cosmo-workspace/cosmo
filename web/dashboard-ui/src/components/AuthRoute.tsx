import React, { ReactElement } from 'react';
import { Navigate, useLocation } from 'react-router-dom';
import { isAdminUser } from '../views/organisms/UserModule';
import { useLogin } from './LoginProvider';

type Props = {
  children: ReactElement;
  admin?: boolean;
}

export const AuthRoute: React.VFC<Props> = ({ children, admin }) => {
  const { loginUser } = useLogin();
  const isAdmin = isAdminUser(loginUser);
  let location = useLocation();

  if (!loginUser) {
    return (<Navigate
      to={{ pathname: '/signin', search: `from=${location.pathname}` }}
      state={{ from: location }}
      replace
    />);
  } else if (admin && !isAdmin) {
    return (<Navigate to='/' replace />);
  } else {
    return children;
  }
}