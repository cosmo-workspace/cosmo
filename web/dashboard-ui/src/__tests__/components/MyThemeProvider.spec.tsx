import { useTheme } from "@emotion/react";
import { useMediaQuery } from "@mui/material";
import { cleanup, renderHook } from "@testing-library/react";
import React from 'react';
import { afterEach, describe, expect, it, MockedFunction, vi } from "vitest";
import { MyThemeProvider } from '../../components/MyThemeProvider';

//--------------------------------------------------
// mock definition
//--------------------------------------------------

vi.mock('@mui/material');

//-----------------------------------------------
// test
//-----------------------------------------------

describe('AuthRoute', () => {

  const useMediaQueryMock = useMediaQuery as MockedFunction<typeof useMediaQuery>;

  afterEach(() => {
    vi.restoreAllMocks();
    cleanup();
  });


  it('normal light', async () => {

    useMediaQueryMock.mockReturnValue(false);
    const { result } = renderHook(() => useTheme(), {
      wrapper: ({ children }) => (<MyThemeProvider>{children}</MyThemeProvider >),
    });
    expect(result.current).toMatchSnapshot();

  });


  it('normal dark', async () => {

    useMediaQueryMock.mockReturnValue(true);
    const { result } = renderHook(() => useTheme(), {
      wrapper: ({ children }) => (<MyThemeProvider>{children}</MyThemeProvider >),
    });
    expect(result.current).toMatchSnapshot();

  });

});
