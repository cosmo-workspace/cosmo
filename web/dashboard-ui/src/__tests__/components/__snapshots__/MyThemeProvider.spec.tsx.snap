// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthRoute > normal dark 1`] = `{}`;

exports[`AuthRoute > normal light 1`] = `{}`;
